src/Build.UnitTests/BackEnd/TaskHostFactory_Tests.cs

```diff
@@ -26,37 +26,114 @@ public TaskHostFactory_Tests(ITestOutputHelper testOutputHelper)
+        [InlineData(false, true)]
+        [InlineData(true, true)]
+        public void TaskNodesDieAfterBuild(bool taskHostFactorySpecified, bool envVariableSpecified)
         {
             using (TestEnvironment env = TestEnvironment.Create())
             {
+
+                string taskFactory = taskHostFactorySpecified ? "TaskHostFactory" : "AssemblyTaskFactory";
                 string pidTaskProject = $@"
 <Project>
-    <UsingTask TaskName=""ProcessIdTask"" AssemblyName=""Microsoft.Build.Engine.UnitTests"" TaskFactory=""TaskHostFactory"" />
+    <UsingTask TaskName=""ProcessIdTask"" AssemblyName=""Microsoft.Build.Engine.UnitTests"" TaskFactory=""{taskFactory}"" />
     <Target Name='AccessPID'>
         <ProcessIdTask>
             <Output PropertyName=""PID"" TaskParameter=""Pid"" />
         </ProcessIdTask>
     </Target>
 </Project>";
                 TransientTestFile project = env.CreateFile("testProject.csproj", pidTaskProject);
+                    
+                if (envVariableSpecified)
+                {
+                    env.SetEnvironmentVariable("MSBUILDFORCEALLTASKSOUTOFPROC", "1");
+                }
                 ProjectInstance projectInstance = new(project.Path);
+
                 projectInstance.Build().ShouldBeTrue();
                 string processId = projectInstance.GetPropertyValue("PID");
                 string.IsNullOrEmpty(processId).ShouldBeFalse();
                 Int32.TryParse(processId, out int pid).ShouldBeTrue();
-                Process.GetCurrentProcess().Id.ShouldNotBe<int>(pid);
-                try
+                Process.GetCurrentProcess().Id.ShouldNotBe(pid);
+
+                if (taskHostFactorySpecified)
+                {
+                    try
+                    {
+                        Process taskHostNode = Process.GetProcessById(pid);
+                        taskHostNode.WaitForExit(3000).ShouldBeTrue($"The executed MSBuild Version: {projectInstance.GetProperty("MSBuildVersion")}");
+                    }
+                    // We expect the TaskHostNode to exit quickly. If it exits before Process.GetProcessById, it will throw an ArgumentException.
+                    catch (ArgumentException e)
+                    {
+                        e.Message.ShouldBe($"Process with an Id of {pid} is not running.");
+                    }
+                }
+                else
                 {
+                    // This is the sidecar TaskHost case - it should persist after build is done. So we need to clean up and kill it ourselves.
                     Process taskHostNode = Process.GetProcessById(pid);
-                    taskHostNode.WaitForExit(2000).ShouldBeTrue();
+                    taskHostNode.WaitForExit(3000).ShouldBeFalse($"The executed MSBuild Version: {projectInstance.GetProperty("MSBuildVersion")}");
+                    taskHostNode.Kill();
+                }
+            }
+        }
+
+        [Fact(Skip = "Environment variable issue")]
```

Comment:
this is a pretty important test to not skip in order to validate the change, no? or do you think your local testing of this was sufficient?